components/authz-service/server/v2/authz.go

```diff
@@ -68,6 +70,27 @@ func (s *authzServer) ProjectsAuthorized(
 		projects = req.ProjectsFilter
 	}
 
+	// this call checks for all allowed projects (without considering deny)
+	if version.Minor == api.Version_V1 && len(req.ProjectsFilter) == 0 {
+		pairs := []*api.Pair{
+			&api.Pair{
+				Resource: req.Resource,
+				Action:   req.Action,
+			},
+		}
+		filterReq := api.FilterAuthorizedPairsReq{
+			Subjects: req.Subjects,
+			Pairs:    pairs,
+		}
+		resp, err := s.FilterAuthorizedProjects(ctx, &filterReq)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		projects = resp.Projects
+	}
```

Comment:
The logic here is getting mind-boggling (I know it was hard to follow to start with! Not blaming this change specifically). It might be better to just break this out into each case (v1.0, v2.0, v2.1 -- or at least v1.0 and v2.x) as separate functions instead of trying to handle them all with this branching. It's getting really, really hard to read through and understand the intentions behind the branching.

So you'd have 

```
func (s *authzServer) ProjectsAuthorized(...) {
         if version.Major == api.Version_V1 {
           projects = s.projectsAuthorizedV1(...)
        }
        ..similar helpers for V2.0 and V2.1

        // This looks like common clean up for every case and can probably stay.
	if stringutils.SliceContains(projectsAuthorized, constants.AllProjectsID) {
		// Engine allows all and we requested all, so signify it as all.
		// This must be different than the requested notion of all,
		// an empty array, because an empty array coming back from the engine means none!
		projectsAuthorized = []string{constants.AllProjectsExternalID}
	}

         return projects
}

func (s *authzServer) v1ProjectsAuthorized(...) {
         projects := []string{}
  	projectsAuthorized, err := s.engine.V2ProjectsAuthorized(ctx,
		engine.Subjects(req.Subjects),
		engine.Action(req.Action),
		engine.Resource(req.Resource),
		engine.ProjectList(projects...))
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())

       if len(projectsAuthorized) > 0 {
         projectsAuthorized = []string{constants.AllProjectsExternalID}
       }
       return projectsAuthorized;
}
```

I think it might be better to duplicate some of the code in `ProjectsAuthorized` at this point for the sake of readability and to make the intentions behind the different branch cases clear. 

**Disclaimer** The above code snippet is just to illustrate the idea and likely contains errors and misses opportunities to clean up the code more.

Reply:
agreed! about to push up some changes similar to what you suggested so i hope that helps!